In [1]:
import tensorflow as tf
import numpy as np
from tensorflow import keras

In [2]:
tf.random.set_seed(42)

In [3]:
epochs = 50
batch_size = 128
verbose = 1
nb_classes = 10
n_hidden = 128
validation_split = 0.2

In [4]:
mnist = keras.datasets.mnist
(X_train, y_train), (X_test, y_test) = mnist.load_data()

In [5]:
print(X_train.shape, X_test.shape)

(60000, 28, 28) (10000, 28, 28)


In [6]:
reshaped = 784

X_train = X_train.reshape(60000, reshaped)
X_test = X_test.reshape(10000, reshaped)

In [7]:
X_train = X_train.astype(float)
X_test = X_test.astype(float)

In [8]:
X_train = X_train / 255
X_test = X_test / 255

In [9]:
y_train = tf.keras.utils.to_categorical(y_train, nb_classes)
y_test = tf.keras.utils.to_categorical(y_test, nb_classes)

In [10]:
model = tf.keras.models.Sequential()
model.add(keras.layers.Dense(n_hidden,
                             input_shape=(reshaped,),
                             activation='relu'))
model.add(keras.layers.Dense(n_hidden,
                             activation='relu'))
model.add(keras.layers.Dense(nb_classes,
                             activation='softmax'))

2022-03-16 16:07:35.776112: I tensorflow/core/platform/cpu_feature_guard.cc:145] This TensorFlow binary is optimized with Intel(R) MKL-DNN to use the following CPU instructions in performance critical operations:  SSE4.1 SSE4.2
To enable them in non-MKL-DNN operations, rebuild TensorFlow with the appropriate compiler flags.
2022-03-16 16:07:35.777156: I tensorflow/core/common_runtime/process_util.cc:115] Creating new thread pool with default inter op setting: 8. Tune using inter_op_parallelism_threads for best performance.


In [11]:
model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense (Dense)                (None, 128)               100480    
_________________________________________________________________
dense_1 (Dense)              (None, 128)               16512     
_________________________________________________________________
dense_2 (Dense)              (None, 10)                1290      
Total params: 118,282
Trainable params: 118,282
Non-trainable params: 0
_________________________________________________________________


In [12]:
model.compile(optimizer='SGD', loss='categorical_crossentropy', metrics=['accuracy'])

In [13]:
model.fit(X_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          verbose=verbose,
          validation_split=validation_split)

Train on 48000 samples, validate on 12000 samples
Epoch 1/50
48000/48000 [==============================] - 3s 61us/sample - loss: 1.5053 - accuracy: 0.6144 - val_loss: 0.7790 - val_accuracy: 0.8322
Epoch 2/50
48000/48000 [==============================] - 2s 48us/sample - loss: 0.6029 - accuracy: 0.8549 - val_loss: 0.4536 - val_accuracy: 0.8832
Epoch 3/50
48000/48000 [==============================] - 2s 49us/sample - loss: 0.4320 - accuracy: 0.8858 - val_loss: 0.3707 - val_accuracy: 0.9003
Epoch 4/50
48000/48000 [==============================] - 2s 46us/sample - loss: 0.3717 - accuracy: 0.8987 - val_loss: 0.3318 - val_accuracy: 0.9075
Epoch 5/50
48000/48000 [==============================] - 2s 46us/sample - loss: 0.3391 - accuracy: 0.9054 - val_loss: 0.3084 - val_accuracy: 0.9138
Epoch 6/50
48000/48000 [==============================] - 2s 46us/sample - loss: 0.3170 - accuracy: 0.9106 - val_loss: 0.2926 - val_accuracy: 0.9183
Epoch 7/50
48000/48000 [==============================] 

In [14]:
test_loss, test_acc = model.evaluate(X_test, y_test)
print('Test Accuracy', test_acc)

10000/1 [===============================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================================

Test Accuracy 0.9645


### Hidden Layers 의 추가로 인해 정확도 상승
- Test Accuracy : from 0.9219 to 0.9645